File: backend/product_grades.py
from dataclasses import dataclass
from typing import Literal


ReloopGrade = Literal["A", "B", "C", "D", "U"]
GradeStatus = Literal[
    "graded",
    "needs_photos",
    "stale",
    "unverified",
]

GRADE_VERSION = "1"
GRADE_DISCLAIMER = (
    "Reloop Grade describes visible condition from listing photos. "
    "It is not an authenticity check or a physical inspection."
)

GRADE_DEFINITIONS = (
    {
        "grade": "A",
        "label": "Like new",
        "description": (
            "No meaningful visible wear in sufficiently clear photos."
        ),
    },
    {
        "grade": "B",
        "label": "Very good",
        "description": (
            "Light visible wear; clean and ready to use."
        ),
    },
    {
        "grade": "C",
        "label": "Good",
        "description": (
            "Noticeable wear or minor flaws, but still usable."
        ),
    },
    {
        "grade": "D",
        "label": "Well worn",
        "description": (
            "Heavy wear, damage, or repair needs are visible."
        ),
    },
    {
        "grade": "U",
        "label": "Unverified",
        "description": (
            "Not enough reliable photo evidence to assign A-D."
        ),
    },
)

GRADE_LABELS = {
    definition["grade"]: definition["label"]
    for definition in GRADE_DEFINITIONS
}


@dataclass(frozen=True)
class GradeDecision:
    grade: ReloopGrade
    status: GradeStatus
    label: str


def derive_reloop_grade(
    *,
    visual_grade: str,
    photo_coverage: str,
    confidence: str,
    seller_condition_consistency: str,
) -> GradeDecision:
    """Convert validated condition evidence into a conservative grade."""

    evidence_is_unsafe = (
        photo_coverage == "limited"
        or confidence == "low"
        or seller_condition_consistency == "review_recommended"
    )

    if evidence_is_unsafe:
        return GradeDecision(
            grade="U",
            status="needs_photos",
            label=GRADE_LABELS["U"],
        )

    grade_by_visual_condition: dict[str, ReloopGrade] = {
        "like_new": "A",
        "good": "B",
        "fair": "C",
        "worn": "D",
    }
    grade = grade_by_visual_condition.get(
        visual_grade,
        "U",
    )

    return GradeDecision(
        grade=grade,
        status=(
            "graded"
            if grade != "U"
            else "needs_photos"
        ),
        label=GRADE_LABELS[grade],
    )


def grade_label(grade: str | None) -> str:
    return GRADE_LABELS.get(
        (grade or "U").upper(),
        GRADE_LABELS["U"],
    )

File: frontend/src/lib/product-grades.ts
export type ReloopGrade = "A" | "B" | "C" | "D" | "U";

export type GradeStatus =
  | "graded"
  | "needs_photos"
  | "stale"
  | "unverified";

export type ProductGradeFields = {
  reloop_grade?: string | null;
  grade_label?: string | null;
  grade_status?: string | null;
  grade_confidence?: string | null;
  grade_summary?: string | null;
  graded_at?: string | null;
};

export type GradeDefinition = {
  grade: ReloopGrade;
  label: string;
  description: string;
  color: string;
  softColor: string;
};

export const GRADE_DEFINITIONS: GradeDefinition[] = [
  {
    grade: "A",
    label: "Like new",
    description: "No meaningful visible wear in clear, useful photos.",
    color: "#236047",
    softColor: "#E7F3EC",
  },
  {
    grade: "B",
    label: "Very good",
    description: "Light visible wear; clean and ready to use.",
    color: "#345E71",
    softColor: "#E8F0F4",
  },
  {
    grade: "C",
    label: "Good",
    description: "Noticeable wear or minor flaws, but still usable.",
    color: "#8A6429",
    softColor: "#F6EEDC",
  },
  {
    grade: "D",
    label: "Well worn",
    description: "Heavy wear, damage, or repair needs are visible.",
    color: "#8A3D32",
    softColor: "#F8E8E5",
  },
  {
    grade: "U",
    label: "Unverified",
    description: "Not enough reliable photo evidence to assign A-D.",
    color: "#655C57",
    softColor: "#EEEAE6",
  },
];

const definitionByGrade = new Map(
  GRADE_DEFINITIONS.map((definition) => [
    definition.grade,
    definition,
  ]),
);

export function normalizeGrade(value?: string | null): ReloopGrade {
  const normalized = (value || "U").trim().toUpperCase();

  if (
    normalized === "A" ||
    normalized === "B" ||
    normalized === "C" ||
    normalized === "D"
  ) {
    return normalized;
  }

  return "U";
}

export function getGradeDefinition(
  value?: string | null,
): GradeDefinition {
  const grade = normalizeGrade(value);
  return definitionByGrade.get(grade) || GRADE_DEFINITIONS[4];
}

export function getGradeDisplay(
  fields: ProductGradeFields,
): GradeDefinition & { status: GradeStatus } {
  const rawStatus = fields.grade_status;
  const status: GradeStatus =
    rawStatus === "graded" ||
    rawStatus === "needs_photos" ||
    rawStatus === "stale"
      ? rawStatus
      : "unverified";

  const grade =
    status === "stale"
      ? "U"
      : normalizeGrade(fields.reloop_grade);
  const definition = getGradeDefinition(grade);

  return {
    ...definition,
    label: fields.grade_label?.trim() || definition.label,
    status,
  };
}
